instrumentation/kotlin-coroutines/src/main/java/io/opentelemetry/instrumentation/auto/kotlincoroutines/KotlinProbeInstrumentation.java

```diff
@@ -0,0 +1,193 @@
+  public Map<? extends ElementMatcher<? super MethodDescription>, String> transformers() {
+    final Map<ElementMatcher<? super MethodDescription>, String> transformers = new HashMap<>();
+    transformers.put(
+        named("probeCoroutineCreated").and(takesArguments(1)),
+        CoroutineCreatedAdvice.class.getName());
+    transformers.put(
+        named("probeCoroutineResumed").and(takesArguments(1)),
+        CoroutineResumedAdvice.class.getName());
+    transformers.put(
+        named("probeCoroutineSuspended").and(takesArguments(1)),
+        CoroutineSuspendedAdvice.class.getName());
+    return transformers;
+  }
+
+  @Override
+  public String[] helperClassNames() {
+    return new String[] {
+      "io.opentelemetry.instrumentation.auto.kotlincoroutines.KotlinProbeInstrumentation$CoroutineWrapper",
+      "io.opentelemetry.instrumentation.auto.kotlincoroutines.KotlinProbeInstrumentation$TraceScopeKey",
+      "io.opentelemetry.instrumentation.auto.kotlincoroutines.KotlinProbeInstrumentation$CoroutineContextWrapper",
+    };
+  }
+
+  public static class CoroutineCreatedAdvice {
+    @Advice.OnMethodExit(onThrowable = Throwable.class, suppress = Throwable.class)
+    public static void exit(
+        @Advice.Return(readOnly = false) kotlin.coroutines.Continuation retVal) {
+      if (!(retVal instanceof CoroutineWrapper)) {
+        retVal = new CoroutineWrapper(retVal);
+      }
+    }
+  }
+
+  public static class CoroutineResumedAdvice {
+    @Advice.OnMethodEnter(suppress = Throwable.class)
+    public static void enter(
+        @Advice.Argument(0) final kotlin.coroutines.Continuation continuation) {
+      continuation.getContext().get(TraceScopeKey.INSTANCE);
+    }
+  }
+
+  public static class CoroutineSuspendedAdvice {
+    @Advice.OnMethodEnter(suppress = Throwable.class)
+    public static void enter(
+        @Advice.Argument(0) final kotlin.coroutines.Continuation continuation) {
+      continuation.getContext().minusKey(TraceScopeKey.INSTANCE);
+    }
+  }
+
+  public static class TraceScopeKey implements CoroutineContext.Key {
+    public static final TraceScopeKey INSTANCE = new TraceScopeKey();
+  }
+
+  public static class CoroutineWrapper implements kotlin.coroutines.Continuation {
+    private final Continuation proxy;
+    private final CoroutineContextWrapper contextWrapper;
+
+    public CoroutineWrapper(Continuation proxy) {
+      this.proxy = proxy;
+      this.contextWrapper = new CoroutineContextWrapper(proxy.getContext());
+    }
+
+    public String toString() {
+      return proxy.toString();
+    }
+
+    @NotNull
+    @Override
+    public CoroutineContext getContext() {
+      return contextWrapper;
+    }
+
+    @Override
+    public void resumeWith(@NotNull Object o) {
+      proxy.resumeWith(o);
+    }
+  }
+
+  public static class CoroutineContextWrapper implements CoroutineContext {
+    private final CoroutineContext proxy;
+    private Context myContext;
+    private Context prevContext;
```

Comment:
As there are a lot of contexts flying around here, I think it would decrease confusion, if these names somehow referenced "tracing" or similar context. To distinguish them from coroutine context.